docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,683 @@
+
+At or shortly after release, we should proactively update the Cosmos SDK to use
+the new API, to remove a disincentive to upgrading.
+
+**Phase 2**: Release v0.37
+
+- During development, we should actively seek out any existing users of the
+  streaming event subscription API and help them migrate.
+- Possibly also: Spend some time writing clients for JS, Rust, et al.
+- Release: Delete the old implementation and all the websocket support code.
+
+> **Discussion point**: Even though the plan is to keep the existing service,
+> we might take the opportunity to restrict the websocket endpoint to _only_
+> the event streaming service, removing the other endpoints. To minimize the
+> disruption for users in the v0.36 cycle, I have decided not to do this for
+> the first phase.
+>
+> If we wind up pushing this design into v0.37, however, we should re-evaulate
+> this partial turn-down of the websocket.
+
+### Future Work
+
+- This design does not immediately address the problem of allowing the client
+  to control which data are reported back for event items. That concern is
+  deferred to future work. However, it would be straightforward to extend the
+  filter and/or the request parameters to allow more control.
+
+- The node currently stores a subset of event data (specifically the block and
+  transaction events) for use in reindexing. While these data are redundant
+  with the event log described in this document, they are not sufficient to
+  cover event subscription, as they omit other event types.
+
+  In the future we should investigate consolidating or removing event data from
+  the state store entirely. For now this issue is out of scope for purposes of
+  updating the RPC API. We may be able to piggyback on the database unification
+  plans (see [RFC 001][rfc001]) to store the event log separately, so its
+  pruning policy does not need to be tied to the block and state stores.
+
+- This design reuses the existing filter query language from the old API.  In
+  the future we may want to use a more structured and/or expressive query.  The
+  Filter object can be extended with more fields as needed to support this.
+
+- Some users have trouble communicating with the RPC service because of
+  configuration problems like improperly-set CORS policies. While this design
+  does not address those issues directly, we might want to revisit how we set
+  policies in the RPC service to make it less susceptible to confusing errors
+  caused by misconfiguration.
+
+---
+## Consequences
+
+- ✅ Reduces the number of transport options for RPC.  Supports [RFC 002][rfc002].
+- ️✅ Removes the primary non-standard use of JSON-RPC.
+- ⛔️ Forces clients to migrate to a different API (eventually).
+- ↕️  API requires clients to poll, but this reduces client state on the server.
+- ↕️  We have to maintain both implementations for a whole release, but this
+  gives clients time to migrate.
+
+---
+## Alternative Approaches
```

Comment:
This use case is more or less exactly the problem that [Server-Sent Events](https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events) (SSE, event sourcing) was meant to solve. I don't see any mention of SSE in this doc — was there a reason it wasn't considered?

Reply:
Perhaps I should have given it a mention. I did look at SSE, but I didn't consider it all that seriously for a couple main reasons:

First and more important, it still requires a persistent connection. One of the goals here was to remove the need for long-running connections, and though the protocol is more resilient to restarts than a vanilla websocket, it still leaves the bookkeeping on the server.

Second, though the SSE plaintext protocol is simple enough that we could pack our data one-line JSON blobs into it, it's yet-another decoding step a client has to understand, for which there is no obvious advantage. (One could argue JSON-RPC wasn't necessarily ideal either, but it's what we've got, and a vocal subset of devs really don't like being forced to use protobuf 🤷🏼‍♀️)

I didn't see an obvious upside to merit pride of place.  I also considered a webhook/websub style "push" type API, but ruled that out for demanding that the server continue to keep track of per-client state, retries, etc., many of the things we're hoping to reduce or eliminate.

Reply:
Okay! If polling was an explicit design goal than indeed SSE isn't a fit.

Reply:
> Okay! If polling was an explicit design goal than indeed SSE isn't a fit.

Less an ideal, perhaps, than a semi-comfortable local minimum in the design space 🙂. Having the consensus node handle event subscription is a blob of accidental complexity that, in retrospect, probably ought to have belonged to the application.  The historical reasons for it are mostly sensible, even if the result is unfortunate.

For me, the practical goal is a simple, working-enough solution that can carry the existing use while we cut away some of the technical debt. It ought to be simple, standard(ish), and enough better than the current thing that the existing user problems are no longer the long pole for RPC issues.